source/common/router/scoped_rds.cc

```diff
@@ -116,44 +137,70 @@ void ScopedRdsConfigSubscription::onConfigUpdate(
       }
       rds.set_route_config_name(scoped_route_config.route_configuration_name());
       ScopedRouteInfoConstSharedPtr scoped_route_info = std::make_shared<ScopedRouteInfo>(
-          std::move(scoped_route_config), srds_config_provider_manager_.createRouteConfigProvider(
-                                              factory_context_, rds, stat_prefix_));
+          std::move(scoped_route_config),
+          route_config_provider_manager_.createRdsRouteConfigProvider(
+              rds, factory_context_, stat_prefix_,
+              (overriding_init_manager == nullptr ? factory_context_.initManager()
+                                                  : *overriding_init_manager)));
       // Detect if there is key conflict between two scopes.
+      ScopedRouteInfoConstSharedPtr prev_scoped_route_info;
       auto iter = scope_name_by_hash_.find(scoped_route_info->scopeKey().hash());
-      if (iter != scope_name_by_hash_.end() && iter->second != scoped_route_info->scopeName()) {
-        throw EnvoyException(
-            fmt::format("scope key conflict found, first scope is '{}', second scope is '{}'",
-                        iter->second, scoped_route_info->scopeName()));
+      if (iter != scope_name_by_hash_.end()) {
+        if (iter->second != scoped_route_info->scopeName()) {
+          throw EnvoyException(
+              fmt::format("scope key conflict found, first scope is '{}', second scope is '{}'",
+                          iter->second, scoped_route_info->scopeName()));
+        }
+        prev_scoped_route_info = scoped_route_map_[scoped_route_info->scopeName()];
       }
       scope_name_by_hash_[scoped_route_info->scopeKey().hash()] = scoped_route_info->scopeName();
       scoped_route_map_[scoped_route_info->scopeName()] = scoped_route_info;
-      applyDeltaConfigUpdate(
-          [scoped_route_info](const ConfigProvider::ConfigConstSharedPtr& config) {
-            auto* thread_local_scoped_config = const_cast<ThreadLocalScopedConfigImpl*>(
-                static_cast<const ThreadLocalScopedConfigImpl*>(config.get()));
+      applyConfigUpdate(
+          [scoped_route_info](
+              ConfigProvider::ConfigConstSharedPtr config) -> ConfigProvider::ConfigConstSharedPtr {
+            auto* thread_local_scoped_config =
+                const_cast<ScopedConfigImpl*>(static_cast<const ScopedConfigImpl*>(config.get()));
 
             thread_local_scoped_config->addOrUpdateRoutingScope(scoped_route_info);
+            return config;
+          },
+          [prev_scoped_route_info]() {
```

Comment:
Can you remind me of the race condition that is protecting against? It's not immediately obvious why this destruction needs to happen on the main thread.

Reply:
so a ScopedRouteInfo contains a RouteConfigProvider, when it's removed from the ScopedConfigImpl, the RCProvider is destructed, we want it to be destructed in main thread as it contains a tls_ member (TLSlot).